Runtime/JSThreadSafeFunction.cs

```diff
@@ -0,0 +1,272 @@
+using System;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using static NodeApi.JSNativeApi.Interop;
+using static NodeApi.JSNativeApi.NodeApiInterop;
+
+namespace NodeApi;
+
+public delegate void JSThreadSafeCallback(JSValue jsFunction, object? functionContext, object? callbackData);
+
+public delegate void JSThreadSafeFinalizeCallback(object? functionContext);
+
+public struct JSThreadSafeFunction
+{
+    private napi_threadsafe_function _tsfn;
+
+    public static explicit operator napi_threadsafe_function(JSThreadSafeFunction function) => function._tsfn;
+    public static implicit operator JSThreadSafeFunction(napi_threadsafe_function tsfn) => new(tsfn);
+
+    private JSThreadSafeFunction(napi_threadsafe_function tsfn) => _tsfn = tsfn;
+
+    // This API may only be called from the main thread.
+    public static unsafe JSThreadSafeFunction New(int maxQueueSize,
+                                                  int initialThreadCount,
+                                                  in JSValue asyncResourceName,
+                                                  in JSValue? jsFunction = null,
+                                                  in JSObject? asyncResource = null,
+                                                  JSThreadSafeFinalizeCallback? finalize = null,
+                                                  object? functionContext = null,
+                                                  JSThreadSafeCallback? jsCaller = null)
+    {
+        FunctionData functionData = new()
+        {
+            FunctionContext = functionContext,
+            Finalize = finalize,
+            JSCaller = jsCaller
+        };
+        GCHandle functionDataHandle = GCHandle.Alloc(functionData);
+        napi_status status = napi_create_threadsafe_function(
+                                 (napi_env)JSValueScope.Current,
+                                 (napi_value)jsFunction,
+                                 (napi_value)(JSValue)(asyncResource is JSObject obj ? obj : new JSObject()),
```

Comment:
> `(asyncResource is JSObject obj ? obj : new JSObject())`

Can be shortened to `(asyncResource ?? new JSObject())`

Reply:
Actually the doc says it's optional, so can it default to `undefined` or a null handle?

Reply:
Checked the code and confirmed that if we pass `null`, then `napi_create_threadsafe_function` creates a new `Object` for us. No need to create it ourselves. I wish I understand what this object is needed for...